File: Tests/Unity/Assets/Tests/UserSystemTests.cs
using System.Collections;
using System.Collections.Generic;

using NUnit.Framework;
using UnityEngine;
using UnityEngine.TestTools;

using Common = Olympus.Foundation.Common;
using Systems = Olympus.Foundation.Systems;
using Services = Olympus.Foundation.Services;


namespace Tests
{
    public class UserSystemTests : TestsBase
    {
        public const string DefaultLoginEmail = "olyservice@rewind.co";
        public const string DefaultLoginPassword = "rcWbfc1MU3Yx";
        public const string SecondTestAccountEmail = "olyservice+test@rewind.co";
        public const string SecondTestAccountPassword = "SsLEW@H=b6J2@7#W";
        public const string SecondTestAccountId = "61f42814b159c6ad996c0c67";

        public const string DefaultDeviceId = "0d011257-d49f-4423-9aa0-06d155e560bb";

        public static IEnumerator LogOut(Systems.UserSystem userSystem)
        {
            var task = userSystem.Logout();

            yield return task.RunAsCoroutine();

            var result = task.Result;
            var resCode = result.GetResultCode();

            Assert.AreEqual(resCode, Services.EResultCode.Success);

            Debug.Log("Logged out");
        }

        public static IEnumerator LogIn(Result<string> outUserId, Systems.UserSystem userSystem, string email = DefaultLoginEmail, string password = DefaultLoginPassword, Services.EResultCode expectedResult = Services.EResultCode.Success)
        {
            var task = userSystem.Login("", email, password);

            yield return task.RunAsCoroutine();

            var result = task.Result;
            var resCode = result.GetResultCode();
            outUserId.Value = result.GetLoginState().UserId;

            Assert.AreEqual(resCode, expectedResult);

            if (resCode == Services.EResultCode.Success)
            {
                PushCleanupFunction(() => LogOut(userSystem));
                Debug.Log($"Logged in (UserId: { outUserId.Value })");
            }
        }

        public static IEnumerator LogInAsGuestWithId(Result<string> outUserId, Systems.UserSystem userSystem, string deviceId = DefaultDeviceId, Services.EResultCode expectedResult = Services.EResultCode.Success, bool pushCleanupFunction = true)
        {
            var task = userSystem.LoginAsGuestWithId(deviceId);

            yield return task.RunAsCoroutine();

            var result = task.Result;
            var resCode = result.GetResultCode();
            outUserId.Value = result.GetLoginState().UserId;

            Assert.AreEqual(resCode, expectedResult);

            if (resCode == Services.EResultCode.Success)
            {
                if (pushCleanupFunction)
                    PushCleanupFunction(() => LogOut(userSystem));

                Debug.Log($"Logged in (UserId: { outUserId.Value })");
            }
        }

        static Dictionary<string, object> ParseJsonObject(string data)
        {
            var json = new Dictionary<string, object>();
            var index = data.IndexOf('{') + 1;

            // Skips whitespace and returns the next character in the string
            char NextChar()
            {
                var c = data[index++];

                while (char.IsWhiteSpace(c))
                    c = data[index++];

                return c;
            }

            // Parses the next string of characters and returns it as the appropriate type
            object ReadValue()
            {
                var c = NextChar();
                object value;

                switch (c)
                {
                    case '{':   // Object
                        {
                            var _start = index - 1;
                            var _scope = 1;

                            while (_scope > 0)
                            {
                                c = data[index++];

                                if (c == '{')
                                    _scope++;
                                else if (c == '}')
                                    _scope--;
                            }

                            var _end = index;
                            var inner = data.Substring(_start, _end - _start);
                            value = ParseJsonObject(inner);
                        }
                        break;
                    case '[':   // Array
                        var list = new List<object>();

                        while (c != ']')
                        {
                            // TODO: FIX THIS! Needs to consume the comma
                            list.Add(ReadValue());
                            c = NextChar();
                        }

                        value = list;
                        break;
                    case '"':   // String
                        {
                            var _start = index;
                            c = data[index++];

                            while (!(c == '"' && data[index - 1] != '\\'))
                                c = data[index++];

                            var _end = index - 1;
                            value = data.Substring(_start, _end - _start);
                        }
                        break;
                    default:
                        {
                            var _start = index - 1;
                            c = data[index++];

                            while (!char.IsWhiteSpace(c))
                            {
                                if (c == ',' || c == '}')
                                {
                                    index--;

                                    break;
                                }

                                c = data[index++];
                            }

                            var _end = index;
                            var inner = data.Substring(_start, _end - _start);

                            if (inner == "true")
                                value = true;
                            else if (inner == "false")
                                value = false;
                            else if (double.TryParse(inner, out var result))
                                value = result;
                            else
                                value = null;
                        }
                        break;
                }

                return value;
            }

            for (; ; )
            {
                var c = NextChar();

                if (c == '}')
                    break;

                if (c == ',')
                    c = NextChar();

                if (c != '"')
                    throw new System.Exception($"Expected '\"'. Got '{c}'.");

                var start = index;
                c = data[index++];

                while (!(c == '"' && data[index - 1] != '\\'))
                    c = data[index++];

                var end = index - 1;
                var key = data.Substring(start, end - start);

                c = NextChar();

                if (c != ':')
                    throw new System.Exception($"Expected ':'. Got '{c}'.");

                json[key] = ReadValue();
            }

            return json;
        }


        [UnityTest, Order(1)]
        public IEnumerator LoginTest()
        {
            GetFoundationSystems(out var userSystem, out _, out _, out _);

            Debug.Log(Olympus.Foundation.OlympusFoundation.GetVersion());

            // Log in
            var userId = new Result<string>();
            yield return LogIn(userId, userSystem);
        }

        [UnityTest, Order(2)]
        public IEnumerator LoginWithTokenTest()
        {
            GetFoundationSystems(out var userSystem, out var spaceSystem, out _, out _);
            string loginToken = "";

            userSystem.OnNewLoginTokenReceived += (s, result) =>
            {
                if (result.GetResultCode() == Services.EResultCode.Success)
                {
                    loginToken = result.GetLoginTokenInfo().RefreshToken;

                    Debug.Log($"New Login token {result.GetLoginTokenInfo().RefreshToken} expires at {result.GetLoginTokenInfo().RefreshExpiryTime}");
                }
            };

            // Log in
            var userId = new Result<string>();
            yield return LogIn(userId, userSystem);

            {
                var task = userSystem.LoginWithToken(userId, loginToken);

                yield return task.RunAsCoroutine();

                var result = task.Result;
                var resCode = result.GetResultCode();
                Assert.AreEqual(result.GetResultCode(), Services.EResultCode.Success);
            }

            var currentLoginState = userSystem.GetLoginState();
            Assert.AreEqual(currentLoginState.State, Systems.ELoginState.LoggedIn);

            //check that we're successfully logged in to CHS by creating a space
            string testSpaceName = GenerateUniqueString("OLY-UNITTEST-SPACE-REWIND");
            string testSpaceDescription = "OLY-UNITTEST-SPACEDESC-REWIND";

            // Create space
            var space = new Result<Systems.Space>();
            yield return SpaceSystemTests.CreateSpace(space, spaceSystem, testSpaceName, testSpaceDescription, Systems.SpaceType.Private);
        }

        [UnityTest, Order(3)]
        public IEnumerator ExchangeKeyTest()
        {
            string key;
            string userId;

            // Use WebClient as HttpClient requires an extra package to be installed
            using (var http = new System.Net.WebClient())
            {
                http.Headers[System.Net.HttpRequestHeader.ContentType] = "text/json";
                var response = http.UploadString(chsEndpointBaseUri + "/mag-user/api/v1/users/login", "{ \"email\": \"olyservice@rewind.co\", \"password\": \"rcWbfc1MU3Yx\", \"deviceId\": \"GoodbyeAndThanksForAllTheFish\" }");
                var json = ParseJsonObject(response);
                userId = json["userId"] as string;
                var authToken = json["accessToken"] as string;

                http.Headers[System.Net.HttpRequestHeader.ContentType] = "text/json";
                http.Headers[System.Net.HttpRequestHeader.Authorization] = "Bearer " + authToken;
                response = http.UploadString(chsEndpointBaseUri + "/mag-user/api/v1/users/onetimekey", $"{{ \"userId\": \"{userId}\", \"deviceId\": \"GoodbyeAndThanksForAllTheFish\" }}");
                json = ParseJsonObject(response);
                key = json["oneTimeKey"] as string;
            }

            GetFoundationSystems(out var userSystem, out _, out _, out _);

            // Log in with OTP
            {
                var task = userSystem.ExchangeKey(userId, key);

                yield return task.RunAsCoroutine();

                var result = task.Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);

                var gotUserID = result.GetLoginState().UserId;

                Assert.AreEqual(userId, gotUserID);
            }
        }

        [UnityTest, Order(4)]
        public IEnumerator GetProfileTest()
        {
            GetFoundationSystems(out var userSystem, out _, out _, out _);

            // Log in
            var userId = new Result<string>();
            yield return LogIn(userId, userSystem);

            // Get profile
            {
                var task = userSystem.GetProfileByUserId(userId);

                yield return task.RunAsCoroutine();

                var result = task.Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
                Assert.AreEqual(result.GetProfile().Email, "olyservice@rewind.co");
            }
        }

        [UnityTest, Order(5)]
        public IEnumerator GetLiteProfilesTest()
        {
            GetFoundationSystems(out var userSystem, out _, out _, out _);

            // Log in
            var userId = new Result<string>();
            yield return LogIn(userId, userSystem);

            // Get profiles
            {
                var userIds = new Common.Array<string>(1);
                userIds[0] = userId;
                var result = userSystem.GetProfilesByUserId(userIds).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
                Assert.AreEqual(result.GetProfiles().Size(), 1);
                Assert.AreEqual(result.GetProfiles()[0].UserId, userId.Value);
            }
        }

        [UnityTest, Order(6)]
        public IEnumerator LoginErrorTest()
        {
            GetFoundationSystems(out var userSystem, out _, out _, out _);

            // Log in with invalid credentials
            var userId = new Result<string>();
            yield return LogIn(userId, userSystem, email: "invalidlogin@rewind.co", password: "", expectedResult: Services.EResultCode.Failed);
        }
    }
}

File: Tests/Web/html/tests/usersystem_tests.js
import { test, assert } from '../test_framework.js';
import { sleep, generateUniqueString, CHS_ENDPOINT_BASE_URI } from '../test_helpers.js';
import { jsArrayToCommonArray } from '../conversion_helpers.js';
import { DEFAULT_LOGIN_EMAIL, DEFAULT_LOGIN_PASSWORD, getProfileByUserId, logIn, logOut } from './usersystem_tests_helpers.js';
import { createSpace } from './spacesystem_tests_helpers.js';

import { Services, Systems } from '../connected_spaces_platform.js';


const ThirdPartyAuthenticationDefines = 
{
    STATE_ID_URL_PARAM : "state=",
    CLIENTID_URL_PARAM : "client_id=",
    SCOPE_URL_PARAM : "scope=",
    REDIRECT_URL_PARAM : "redirect_uri=",
    INVALID_URL_PARAM_VALUE : "N/A"
}


function validateThirdPartyAuthoriseURL(authoriseURL, redirectURL)
{
    assert.areNotEqual(authoriseURL, "");
    assert.areNotEqual(authoriseURL, "error");

    const tokens = authoriseURL.split(/[&?]+/); // Regex looking for & and ? characters, multiple times 

    let stateId = ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE;
    let clientId = ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE;
    let scope = ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE;
    let retrievedRedirectURL = ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE;

    for (let i = 0; i < tokens.length; i++)
    {
        let URLElement = tokens[i];

        if(URLElement.startsWith(ThirdPartyAuthenticationDefines.STATE_ID_URL_PARAM))
            stateId = URLElement.substring(ThirdPartyAuthenticationDefines.STATE_ID_URL_PARAM.length);
        else if (URLElement.startsWith(ThirdPartyAuthenticationDefines.CLIENTID_URL_PARAM))
            clientId = URLElement.substring(ThirdPartyAuthenticationDefines.CLIENTID_URL_PARAM.length);
        else if(URLElement.startsWith(ThirdPartyAuthenticationDefines.SCOPE_URL_PARAM))
            scope = URLElement.substring(ThirdPartyAuthenticationDefines.SCOPE_URL_PARAM.length);
        else if(URLElement.startsWith(ThirdPartyAuthenticationDefines.REDIRECT_URL_PARAM))
            retrievedRedirectURL = URLElement.substring(ThirdPartyAuthenticationDefines.REDIRECT_URL_PARAM.length);
    }

    // Validate that the following contain something that potentially makes sense
    assert.areNotEqual(stateId, ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE);
    assert.areNotEqual(clientId, ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE);
    assert.areNotEqual(scope, ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE);
    assert.areNotEqual(retrievedRedirectURL, ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE);

    assert.isGreaterThan(stateId.length, 10);
    assert.isGreaterThan(clientId.length, 10);
    assert.isGreaterOrEqualThan(scope.length, 5);
    assert.areEqual(retrievedRedirectURL, redirectURL);
}


async function getFullProfileByUserId(userSystem, userId)
{
    const result = await userSystem.getProfileByUserId(userId);
    
    assert.succeeded(result);

    const profile = result.getProfile();
    result.delete();

    return profile;
}


test('UserSystemTests', 'LoginTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in
    await logIn(userSystem);
});

test('UserSystemTests', 'LoginTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in
    await logIn(userSystem);
});

test('UserSystemTests', 'FalseAgeVerificationLoginTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in with false age verification
    {
        const result = await userSystem.login('', DEFAULT_LOGIN_EMAIL, DEFAULT_LOGIN_PASSWORD,false);
    
        assert.failed(result);

        result.delete();
    }

    // Log in with true age verification
    {
        const result = await userSystem.login('', DEFAULT_LOGIN_EMAIL, DEFAULT_LOGIN_PASSWORD,true);
    
        assert.succeeded(result);

        result.delete();
    }

});


test('UserSystemTests', 'LoginWithTokenTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();
    const spaceSystem = systemsManager.getSpaceSystem();

    let loginToken = "";
    let loginTokenAvailable = false;

    userSystem.setNewLoginTokenReceivedCallback((result) => {
        const resCode = result.getResultCode();

        assert.succeeded(result);

        if (resCode === Services.EResultCode.Success) {
            const tokenInfo = result.getLoginTokenInfo();
            const tokenExpiry = tokenInfo.refreshExpiryTime;
            loginToken = tokenInfo.refreshToken;
            tokenInfo.delete();

            loginTokenAvailable = true;

            console.debug(`New login token ${loginToken} expires at ${tokenExpiry}`);
        }

        result.delete();
    });

    // Log in
    const userId = await logIn(userSystem);

    let waitForTestTimeoutCountMs = 0;
    const WAIT_FOR_TEST_TIMEOUT_LIMIT_MS = 1000;

    while (!loginTokenAvailable && (waitForTestTimeoutCountMs < WAIT_FOR_TEST_TIMEOUT_LIMIT_MS)) {
        await sleep(50);
        waitForTestTimeoutCountMs += 50;
    }

    loginTokenAvailable = false;
    waitForTestTimeoutCountMs = 0;

    // Log in with refresh token
    {
        const result = await userSystem.loginWithToken(userId, loginToken);
        
        assert.succeeded(result);

        result.delete();
    }

    while (!loginTokenAvailable && waitForTestTimeoutCountMs < WAIT_FOR_TEST_TIMEOUT_LIMIT_MS) {
        await sleep(50);
        waitForTestTimeoutCountMs += 50;
    }
    
    const currentLoginState = userSystem.getLoginState();
    
    assert.areEqual(currentLoginState.state, Systems.ELoginState.LoggedIn);
    
    currentLoginState.delete();

    // Create space to verify that we successfully logged in
    const spaceName = generateUniqueString('CSP-TESTS-WASM-SPACE');
    const spaceDescription = 'CSP-TESTS-WASM-SPACEDESC';

    await createSpace(spaceSystem, spaceName, spaceDescription, Systems.SpaceAttributes.Private);
});


test('UserSystemTests', 'ExchangeKeyTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    let response = await fetch(CHS_ENDPOINT_BASE_URI + '/mag-user/api/v1/users/login', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            email: DEFAULT_LOGIN_EMAIL,
            password: DEFAULT_LOGIN_PASSWORD,
            deviceId: 'GoodbyeAndThanksForAllTheFish'
        })
    });

    let json = await response.json();
    const userId = json.userId;
    const accessToken = json.accessToken;

    console.debug(`Logged in via fetch (UserId: ${userId})`);

    response = await fetch(CHS_ENDPOINT_BASE_URI + '/mag-user/api/v1/users/onetimekey', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${accessToken}`
        },
        body: JSON.stringify({
            userId: userId,
            deviceId: 'GoodbyeAndThanksForAllTheFish'
        })
    });

    json = await response.json();
    const oneTimeKey = json.oneTimeKey;

    console.debug(`Got OTK via fetch (${oneTimeKey})`);

    // Log in with OTP
    {
        const result = await userSystem.exchangeKey(userId, oneTimeKey);
        
        assert.succeeded(result);

        const loginState = result.getLoginState();
        result.delete();

        const receivedUserId = loginState.userId;
        loginState.delete();

        assert.areEqual(receivedUserId, userId);

        console.debug(`Exchanged key (UserId: ${receivedUserId})`);
    }

    // Log out
    await logOut(userSystem);
});


test('UserSystemTests', 'LoginErrorTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in with invalid credentials
    await logIn(userSystem, 'invalidlogin@magnopus.com', 'notarealpassword', Services.EResultCode.Failed);

    // Log in
    await logIn(userSystem);
});


test('UserSystemTests', 'UpdateDisplayNameTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in
    const userId = await logIn(userSystem);
    const newDisplayName = generateUniqueString('CSP-TESTS-WASM-NAME').substring(0, 24);

    // Update display name
    {
        const result = await userSystem.updateUserDisplayName(userId, newDisplayName);
        
        assert.succeeded(result);

        result.delete();
    }

    // Retrieve user profile and verify new display name
    const profile = await getProfileByUserId(userSystem, userId);

    assert.areEqual(profile.userId, userId);
    assert.areEqual(profile.displayName, newDisplayName);

    profile.delete();
});


test('UserSystemTests', 'UpdateDisplayNameTestWithSpaces', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in
    const userId = await logIn(userSystem);
    const newDisplayName = generateUniqueString('OLY TESTS WASM NAME WITH SPACES').substring(0, 24);

    // Update display name
    {
        const result = await userSystem.updateUserDisplayName(userId, newDisplayName);
        
        assert.succeeded(result);

        result.delete();
    }

    // Retrieve user profile and verify new display name
    const profile = await getProfileByUserId(userSystem, userId);

    assert.areEqual(profile.userId, userId);
    assert.areEqual(profile.displayName, newDisplayName);

    profile.delete();
});


test('UserSystemTests', 'UpdateDisplayNameTestWithSymbols', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in
    const userId = await logIn(userSystem);
    const newDisplayName = generateUniqueString('NAME SYMBOLS *&^%%$@~#!;?.,{}[]()=').substring(0, 24);

    // Update display name
    {
        const result = await userSystem.updateUserDisplayName(userId, newDisplayName);
        
        assert.succeeded(result);

        result.delete();
    }

    // Retrieve user profile and verify new display name
    const profile = await getProfileByUserId(userSystem, userId);

    assert.areEqual(profile.userId, userId);
    assert.areEqual(profile.displayName, newDisplayName);

    profile.delete();
});

test('UserSystemTests', 'UpdateDisplayNameTestWithTooLongName', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in
    const userId = await logIn(userSystem);

    const newDisplayName = generateUniqueString('OLY WASM{}[()= LONG NAME MORE THAN 24 CHARACTERS');

    // Update display name
    {
        const result = await userSystem.updateUserDisplayName(userId, newDisplayName);
        
        assert.failed(result);

        result.delete();
    }

    // Retrieve user profile and verify new display name
    const profile = await getProfileByUserId(userSystem, userId);

    profile.delete();
});

test('UserSystemTests', 'PingTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Ping UserSystem CHS services
    const result = await userSystem.ping();   
    
    assert.succeeded(result);

    result.delete();
});


test('UserSystemTests', 'CreateUserTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();
    const settingsSystem = systemsManager.getSettingsSystem();

    const newUsername = generateUniqueString("OLY-TEST-NAME");
    const newDisplayName = "OLY-TEST-DISPLAY";
    const newEmail = generateUniqueString('testnopus.') + "@magnopus.com";
    const newPassword = generateUniqueString('Password');

    let createdUserId;
    
    // Test user creation
    {
        const result = await userSystem.createUser(newUsername, newDisplayName, newEmail, newPassword, true, null, null);
        
        assert.succeeded(result);

        const profile = result.getProfile();
        result.delete();
        
        createdUserId = profile.userId;

        console.debug(`Created user (UserId: ${profile.userId})`);

        assert.areEqual(profile.userName, newUsername);
        assert.areEqual(profile.displayName, newDisplayName);
        assert.areEqual(profile.email, newEmail);

        profile.delete();
    }

    // Log in
    await logIn(userSystem);

    // Verify that newsletter preference was set
    {
        const result = await settingsSystem.getNewsletterStatus(createdUserId);

        assert.succeeded(result);
        assert.isTrue(result.getValue());
    }

    // Test lite profile
    {
        let userIds = [ createdUserId ];
        let _userIds = jsArrayToCommonArray(userIds, String);
        const result = await userSystem.getProfilesByUserId(_userIds);

        _userIds.delete();

        assert.succeeded(result);

        const profiles = result.getProfiles();
        result.delete();

        assert.areEqual(profiles.size(), 1);

        const profile = profiles.get(0);
        profiles.delete();

        assert.areEqual(profile.userId, createdUserId);
        assert.areEqual(profile.displayName, newDisplayName);

        profile.delete();
    }

    // Test full profile
    {
        const result = await userSystem.getProfileByUserId(createdUserId);

        assert.succeeded(result);

        const profile = result.getProfile();
        result.delete();

        assert.areEqual(profile.userId, createdUserId);
        assert.areEqual(profile.userName, newUsername);
        assert.areEqual(profile.displayName, newDisplayName);
        assert.areEqual(profile.email, newEmail);

        profile.delete();
    }

    // Cleanup
    await userSystem.deleteUser(createdUserId);

    console.debug(`Deleted user (UserId: ${createdUserId})`);
});


test('UserSystemTests', 'GetAuthoriseURLForGoogleTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    let redirectURL = "https://odev.magnoverse.space/oauth";
    
    // Retrieve Authorise URL for Google
    const result = await userSystem.getThirdPartyProviderAuthoriseURL(Systems.EThirdPartyAuthenticationProviders.Google, redirectURL);

    assert.succeeded(result);

    let authoriseURL = result.getValue();
    result.delete();

    validateThirdPartyAuthoriseURL(authoriseURL, redirectURL);
});


test('UserSystemTests', 'GetAuthoriseURLForDiscordTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    let redirectURL = "https://odev.magnoverse.space/oauth";
    
    // Retrieve Authorise URL for Discord
    const result = await userSystem.getThirdPartyProviderAuthoriseURL(Systems.EThirdPartyAuthenticationProviders.Discord, redirectURL);

    assert.succeeded(result);

    let authoriseURL = result.getValue();
    result.delete();

    validateThirdPartyAuthoriseURL(authoriseURL, redirectURL);
});

test('UserSystemTests', 'GetAuthoriseURLForAppleTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    let redirectURL = "https://odev.magnoverse.space/oauth";
    
    // Retrieve Authorise URL for Discord
    const result = await userSystem.getThirdPartyProviderAuthoriseURL(Systems.EThirdPartyAuthenticationProviders.Apple, redirectURL);

    assert.succeeded(result);

    let authoriseURL = result.getValue();
    result.delete();

    validateThirdPartyAuthoriseURL(authoriseURL, redirectURL);
});
// As the following two tests require manual actions explained inside, they are currently disabled
// only these WASM tests would be able to have a end-to-end testing flow using Selenium for the URL redirects. 
// The Selenium work involved is part of a separate ticket
/*

test('UserSystemTests', 'GoogleLogInTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    let redirectURL = "https://odev.magnoverse.space/oauth";
    
    // Retrieve Authorise URL for Google
    {
        const result = await userSystem.getThirdPartyProviderAuthoriseURL(Systems.EThirdPartyAuthenticationProviders.Google, redirectURL);
        
        assert.succeeded(result);

        let authoriseURL = result.getValue();
        result.delete();

        let tokens = authoriseURL.split("&");
        let stateId = ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE;

        for (let i = 0; i < tokens.length; i++)
        {
            let URLElement = tokens[i];

            if(URLElement.startsWith(ThirdPartyAuthenticationDefines.STATE_ID_URL_PARAM))
            {
                stateId = URLElement.substring(ThirdPartyAuthenticationDefines.STATE_ID_URL_PARAM.length);
                break;
            }
        }

        console.log("AuthoriseURL: ", authoriseURL);
    }

    // 1. Set a breakpoint on the next line before reading from the file
    // 2. Navigate to the AuthoriseURL in a browser, but make sure that for the Third party account you're using there's already a created CHS account (same email address)
    // 3. Get the "code" param value from the response URL 
    // 4. Drop the token it in the file below
    // 5. Due to the fact that the browser caches the file content you will have to "Empty cache and and hard refresh" in Chrome now by right click-ing on the Refresh button and select the option.
    // 6. Rerun the same test and when it gets again here to read the file content your new token will be ready for use!

    const LocalFilePath = "/assets/third_party_auth_token.txt";
    const FilePath = window.location.host + LocalFilePath;
    let response = await fetch("http://" + FilePath,{
        method: 'get',
    });

    let googleToken = await response.text();

    {
        const result = await userSystem.loginToThirdPartyAuthenticationProvider(googleToken, stateId);

        assert.succeeded(result);

        const loginState = result.getLoginState();
        result.delete();

        const userId = loginState.userId;
        loginState.delete();

        const fullProfile = await getFullProfileByUserId(userSystem, userId);
        console.log(fullProfile);

        fullProfile.delete();
    }
});


test('UserSystemTests', 'DiscordLogInTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    let redirectURL = "https://odev.magnoverse.space/oauth";

    // Retrieve Authorise URL for Discord
    {
        const result = await userSystem.getThirdPartyProviderAuthoriseURL(Systems.EThirdPartyAuthenticationProviders.Discord, redirectURL);

        assert.succeeded(result);

        let authoriseURL = result.getValue();
        result.delete();

        let tokens = authoriseURL.split("&");
        let stateId = ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE;

        for (let i = 0; i < tokens.length; i++)
        {
            let URLElement = tokens[i];

            if(URLElement.startsWith(ThirdPartyAuthenticationDefines.STATE_ID_URL_PARAM))
            {
                stateId = URLElement.substring(ThirdPartyAuthenticationDefines.STATE_ID_URL_PARAM.length);
                break;
            }
        }

        console.log("AuthoriseURL: ", authoriseURL);
    }

    // 1. Set a breakpoint on the next line before reading from the file
    // 2. Navigate to the AuthoriseURL in a browser, but make sure that for the Third party account you're using there's already a created CHS account (same email address)
    // 3. Get the "code" param value from the response URL 
    // 4. Drop the token it in the file below
    // 5. Due to the fact that the browser caches the file content you will have to "Empty cache and and hard refresh" in Chrome now by right click-ing on the Refresh button and select the option.
    // 6. Rerun the same test and when it gets again here to read the file content your new token will be ready for use!
    
    const LocalFilePath = "/assets/third_party_auth_token.txt";
    const FilePath = window.location.host + LocalFilePath;
    let response = await fetch("http://" + FilePath,{
        method: 'get',
    });

    let discordToken = await response.text();

    {
        const result = await userSystem.loginToThirdPartyAuthenticationProvider(discordToken, stateId);
        assert.succeeded(result);

        const loginState = result.getLoginState();
        result.delete();

        const userId = loginState.userId;
        loginState.delete();

        const fullProfile = await getFullProfileByUserId(userSystem, userId);
        console.log(fullProfile);

        fullProfile.delete();
    }
});

test('UserSystemTests', 'AppleLogInTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    let redirectURL = "https://example-app.com/redirect";
    let stateId = ThirdPartyAuthenticationDefines.INVALID_URL_PARAM_VALUE;

    // Retrieve Authorise URL for Discord
    {
        const result = await userSystem.getThirdPartyProviderAuthoriseURL(Systems.EThirdPartyAuthenticationProviders.Apple, redirectURL);
        console.log(result.getResponseBody());
        assert.succeeded(result);

        let authoriseURL = result.getValue();
        result.delete();

        let tokens = authoriseURL.split("&");

        for (let i = 0; i < tokens.length; i++)
        {
            let URLElement = tokens[i];

            if(URLElement.startsWith(ThirdPartyAuthenticationDefines.STATE_ID_URL_PARAM))
            {
                stateId = URLElement.substring(ThirdPartyAuthenticationDefines.STATE_ID_URL_PARAM.length);
                break;
            }
        }

        console.log("AuthoriseURL: ", authoriseURL);
    }

    // 1. Set a breakpoint on the next line before reading from the file
    // 2. Navigate to the AuthoriseURL in a browser, but make sure that for the Third party account you're using there's already a created CHS account (same email address)
    // 3. Get the "code" param value from the response URL 
    // 4. Drop the token it in the file below
    // 5. Due to the fact that the browser caches the file content you will have to "Empty cache and and hard refresh" in Chrome now by right click-ing on the Refresh button and select the option.
    // 6. Rerun the same test and when it gets again here to read the file content your new token will be ready for use!
    
    const LocalFilePath = "/assets/third_party_auth_token.txt";
    const FilePath = window.location.host + LocalFilePath;
    let response = await fetch("http://" + FilePath,{
        method: 'get',
    });

    let AppleToken = await response.text();

    {
        const result = await userSystem.loginToThirdPartyAuthenticationProvider(AppleToken, stateId);
        assert.succeeded(result);

        const loginState = result.getLoginState();
        result.delete();

        const userId = loginState.userId;
        loginState.delete();

        const fullProfile = await getFullProfileByUserId(userSystem, userId);
        console.log(fullProfile);

        fullProfile.delete();
    }
});

// Disabled by default as it requires superuser access to delete the account

test('UserSystemTests', 'CreateUserEmptyUsernameDisplaynameTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    const newEmail = generateUniqueString('testnopus.') + "@magnopus.com";
    const newPassword = generateUniqueString('Password');

    let userId = "";
    
    // Test user creation
    {
        const result = await userSystem.createUser(null, null, newEmail, newPassword, false, null);
        
        assert.succeeded(result);

        const profile = result.getProfile();
        result.delete();

        userId = profile.userId;

        assert.isTrue(profile.userName.length === 0);
        assert.isFalse(profile.displayName.length === 0);
        assert.areEqual(profile.email, newEmail);

        profile.delete();
    }

    // Log in
    await logIn(userSystem);

    // Test lite profile
    {
        const userIds = [ userId ];
        const _userIds = jsArrayToCommonArray(userIds, String);

        const result = await userSystem.getProfilesByUserId(_userIds);
        _userIds.delete();

        assert.succeeded(result);

        const profiles = result.getProfiles();
        result.delete();

        const profile = profiles.get(0);

        assert.areEqual(profile.userId, userId);
        assert.isFalse(profile.displayName.length === 0);

        profile.delete();
    }

    // Test full profile
    {
        const result = await userSystem.getProfileByUserId(userId);

        assert.succeeded(result);

        const profile = result.getProfile();

        assert.areEqual(profile.userId, userId);

        assert.isTrue(profile.userName.length === 0)
        assert.isFalse(profile.displayName.length === 0);
        assert.areEqual(profile.email, newEmail);

        profile.delete();
    }

    // Cleanup
    await userSystem.deleteUser(userId);
});
*/

test('UserSystemTests', 'GetAgoraUserTokenTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    // Log in
    await logIn(userSystem);

    /*
	  Setup token params 
	  For this test, it's exeptable to use 0 for the UserId and ChannelId
	  This is because the endpoint is just using an algorithm to generate the token
	  So no valid Ids are needed for verification
	*/
    const params = Systems.AgoraUserTokenParams.create();
    params.agoraUserId = "0";
    params.channelName = "0";
    params.lifespan = 10000;
    params.shareAudio = true;
    params.shareScreen = false;
    params.shareVideo = false;
    params.readOnly = false;

    // Get token
    const result = await userSystem.getAgoraUserToken(params);

    assert.succeeded(result);
    assert.areNotEqual(result.getUserToken(), "");
});

test('UserSystemTests', 'GetGuestProfileTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();

    const id = await logIn(userSystem);

    // Uncomment below to run this as a test of the access token refresh
    // await new Promise(r => setTimeout(r, 40 * 60 * 1000));

    getFullProfileByUserId(userSystem, id);
});